File: backend/app/models/roles.py
"""DTOs for all role-related resources: roles, permissions, tools, and the
role ↔ permission / role ↔ tool grants."""

from __future__ import annotations

from pydantic import BaseModel

from app.db.models.enums import RiskLevel
from app.models.base import ORMModel



class PermissionRead(ORMModel):
    id: str
    name: str
    description: str | None = None


class PermissionCreate(BaseModel):
    name: str
    description: str | None = None




class ToolRead(ORMModel):
    id: str
    name: str
    provider: str | None = None
    description: str | None = None


class ToolCreate(BaseModel):
    name: str
    provider: str | None = None
    description: str | None = None


class RoleToolAccess(ORMModel):
    """A tool granted to a role, with its enabled flag (role_tools row)."""

    tool: ToolRead
    enabled: bool = True




class RoleBase(BaseModel):
    name: str
    description: str | None = None
    personality: dict | None = None
    tone: str | None = None
    risk_level: RiskLevel = RiskLevel.MEDIUM
    is_active: bool = True


class RoleCreate(RoleBase):
    permission_ids: list[str] = []
    tool_ids: list[str] = []


class RoleUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    personality: dict | None = None
    tone: str | None = None
    risk_level: RiskLevel | None = None
    is_active: bool | None = None
    permission_ids: list[str] | None = None
    tool_ids: list[str] | None = None


class RoleRead(RoleBase, ORMModel):
    id: str
    permissions: list[PermissionRead] = []
    tools: list[RoleToolAccess] = []

File: apps/web/src/lib/roles.ts
/** Roles/permissions/tools API calls. Types mirror backend/app/models/roles.py. */
import { apiFetch } from "./api-client";

export type RiskLevel = "low" | "medium" | "high";

export interface PermissionRead {
  id: string;
  name: string;
  description: string | null;
}

export interface ToolRead {
  id: string;
  name: string;
  provider: string | null;
  description: string | null;
}

export interface RoleToolAccess {
  tool: ToolRead;
  enabled: boolean;
}

/**
 * Free-form role personality/config. The backend only guarantees this is a
 * JSON object — this is our own agreed-upon shape for it, written and read
 * by this app only (no server-side validation of these specific fields).
 */
export interface RolePersonality {
  summary?: string;
  traits?: string[];
  systemPromptPreview?: string;
  responseStyle?: "concise" | "balanced" | "detailed";
  behaviorPreferences?: string[];
}

export interface RoleRead {
  id: string;
  name: string;
  description: string | null;
  personality: RolePersonality | null;
  tone: string | null;
  risk_level: RiskLevel;
  is_active: boolean;
  permissions: PermissionRead[];
  tools: RoleToolAccess[];
}

export interface RoleCreate {
  name: string;
  description?: string | null;
  personality?: RolePersonality | null;
  tone?: string | null;
  risk_level?: RiskLevel;
  is_active?: boolean;
  permission_ids?: string[];
  tool_ids?: string[];
}

export type RoleUpdate = Partial<RoleCreate>;

/** First word of the role name — the backend has no separate "short name". */
export function shortName(name: string): string {
  return name.split(" ")[0] ?? name;
}

/** Category label derived from a dotted permission name, e.g. "portfolio.read" -> "Portfolio". */
export function permissionCategory(permissionName: string): string {
  const [prefix] = permissionName.split(".");
  const label = prefix || permissionName;
  return label.charAt(0).toUpperCase() + label.slice(1);
}

export function listRoles(): Promise<RoleRead[]> {
  return apiFetch<RoleRead[]>("/api/roles");
}

export function getRole(roleId: string): Promise<RoleRead> {
  return apiFetch<RoleRead>(`/api/roles/${roleId}`);
}

export function createRole(input: RoleCreate): Promise<RoleRead> {
  return apiFetch<RoleRead>("/api/roles", { method: "POST", json: input });
}

export function updateRole(roleId: string, input: RoleUpdate): Promise<RoleRead> {
  return apiFetch<RoleRead>(`/api/roles/${roleId}`, { method: "PATCH", json: input });
}

export function listPermissions(): Promise<PermissionRead[]> {
  return apiFetch<PermissionRead[]>("/api/permissions");
}

export function listTools(): Promise<ToolRead[]> {
  return apiFetch<ToolRead[]>("/api/tools");
}
